src/envoy/http/authn/http_filter_test.cc

```diff
@@ -0,0 +1,144 @@
+/* Copyright 2018 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "src/envoy/http/authn/http_filter.h"
+#include "authentication/v1alpha1/policy.pb.h"
+#include "common/http/header_map_impl.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "src/envoy/http/authn/authenticator_base.h"
+#include "test/mocks/http/mocks.h"
+
+using Envoy::Http::Istio::AuthN::AuthenticatorBase;
+using Envoy::Http::Istio::AuthN::FilterContext;
+using testing::_;
+using testing::Invoke;
+using testing::NiceMock;
+using testing::StrictMock;
+
+namespace Envoy {
+namespace Http {
+namespace {
+
+// Create a fake authenticator for test. This authenticator do nothing except
+// making done callback with failure.
+std::unique_ptr<AuthenticatorBase> createAlwaysFailAuthenticator(
```

Comment:
I think it is better to use MockAuthetncatorBase and use EXPECT_CALL and invoke to custom Mocked run() call. for these two classes,  createAlwaysFailAuthenticatior and createAlwaysPassAuthenticator

Reply:
I explained earlier, in test, we don't have access to the object returned by these functions, so cannot place an EXPECT_CALL

Reply:
MockAuthenticatorBase can be a member of test class.